work_queue/src/work_queue.c

```diff
@@ -851,31 +847,76 @@ static int get_transfer_wait_time(struct work_queue *q, struct work_queue_worker
 	return timeout;
 }
 
-static void remove_factory(struct work_queue_factory_info *f)
+static int factory_trim_workers(struct work_queue *q, struct work_queue_factory_info *f)
+{
+	// Iterate through all workers and shut idle ones down
+	char *factory_name = f->name;
+	struct work_queue_worker *w;
+	char *key;
+	int i = 0;
+
+	hash_table_firstkey(q->worker_table);
+	while( f->connected_workers > f->max_workers &&
+			hash_table_nextkey(q->worker_table, &key, (void **) &w) ) {
+		if( !strcmp(factory_name, w->factory_name) &&
+			!itable_size(w->current_tasks) ) {
+			shut_down_worker(q, w);
+			i++;
+		}
+	}
+
+	// Trimmed some but not enough workers.
+	if (f->connected_workers > f->max_workers) f->trimmed = 1;
+	debug(D_WQ, "Trimmed %d workers from %s", i, factory_name);
+	return i;
+}
+
+static struct work_queue_factory_info* create_factory_info(struct work_queue *q, const char *name)
```

Comment:
`static struct work_queue_factory_info *create_factory_info`
Rest of the code usually puts the * next to the name of function.